### ISSUE TITLE
MNIST 使用说明

### ISSUE THREAD

**Issue by nonocast** [nonocast/me#220]
# MNIST

官网:  [MNIST handwritten digit database, Yann LeCun, Corinna Cortes and Chris Burges](http://yann.lecun.com/exdb/mnist/)

> The MNIST database of handwritten digits, available from this page, has a training set of 60,000 examples, and a test set of 10,000 examples. It is a subset of a larger set available from NIST. The digits have been size-normalized and centered in a fixed-size image.
It is a good database for people who want to try learning techniques and pattern recognition methods on real-world data while spending minimal efforts on preprocessing and formatting.

简单查了一下background:
- NIST: National Institute of Standards and Technology, 美国国家标准与技术研究院，原NBS(National Bureau of Standards, 美国国家标准局) 可以通过 [From NBS to NIST | NIST](https://www.nist.gov/history/nist-100-foundations-progress/nbs-nist) 了解
- NIST 公开发布了很多[special database](https://www.nist.gov/srd/shop/special-database-catalog), 本意只是来自人口普查局的表单照片，然后Yann LeCun以这个为基础，选取了一部分数据做了统一尺寸以及居中等处理形成了MNIST Database of handwritten digits, 第一个M网上有mixed和modified两个解释，这里更多的是mixed, SD-3来自人口普查局，SD-1来自高中生，SD-3更为清晰，而MNIST则是将两个数据库打混
- 有兴趣可以看下[NIST Special Database 19 | NIST](https://www.nist.gov/srd/nist-special-database-19), 下载下来1.03G，其中就是png，然后这帮人想png太大，所以用vector来抽象，就有了idx
- [train-images-idx3-ubyte.gz](http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz):  training set images (9912422 bytes) 
- [train-labels-idx1-ubyte.gz](http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz):  training set labels (28881 bytes) 
- [t10k-images-idx3-ubyte.gz](http://yann.lecun.com/exdb/mnist/t10k-images-idx3-ubyte.gz):   test set images (1648877 bytes) 
- [t10k-labels-idx1-ubyte.gz](http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz):   test set labels (4542 bytes)

## IDX

官网往下就能看到`FILE FORMATS FOR THE MNIST DATABASE`

> The data is stored in a very simple file format designed for storing vectors and multidimensional matrices.

idx是基于binary格式, 布局如下:
```
magic number (32 bit)
size in dimension 0 (32 bit)
size in dimension 1 (32 bit)
size in dimension 2 (32 bit)
.....
size in dimension N  (32 bit)
data
```

TRAINING SET LABEL FILE (train-labels-idx1-ubyte):
```
[offset] [type]          [value]          [description] 
0000     32 bit integer  0x00000801(2049) magic number (MSB first) 
0004     32 bit integer  60000            number of items 
0008     unsigned byte   ??               label 
0009     unsigned byte   ??               label 
........ 
xxxx     unsigned byte   ??               label

The labels values are 0 to 9. 即对应数字0-9
```

先说明一下magic number (32bit)的构成，

```
The magic number is an integer (MSB first). The first 2 bytes are always 0.

The third byte codes the type of the data: 
0x08: unsigned byte 
0x09: signed byte 
0x0B: short (2 bytes) 
0x0C: int (4 bytes) 
0x0D: float (4 bytes) 
0x0E: double (8 bytes)

The 4-th byte codes the number of dimensions of the vector/matrix: 1 for vectors, 2 for matrices....
```

0x00000801，16个0起始，08表示each dimension都是unsigned byte，即data每个unit是1个ubyte，最后01表示1个dimension，data中第1个字节对应索引0，第100个字节对应索引99.

假设我们有4张图片分别对应1,3,5,7这4个数字，则字节流应该是`0x00000801 0x00000004 0x01030507`

根据这个逻辑再来看TRAINING SET IMAGE FILE (train-images-idx3-ubyte):
```
[offset] [type]          [value]          [description] 
0000     32 bit integer  0x00000803(2051) magic number 
0004     32 bit integer  60000            number of images 
0008     32 bit integer  28               number of rows 
0012     32 bit integer  28               number of columns 
0016     unsigned byte   ??               pixel 
0017     unsigned byte   ??               pixel 
........ 
xxxx     unsigned byte   ??               pixel

Pixels are organized row-wise. Pixel values are 0 to 255. 0 means background (white), 255 means foreground (black). 
```

首先文件名中idx3-ubyte，对应表示3个dimension, data的单元类型为ubyte, 3维分别是图片数量，单张图片的row像素数量和单张图片column像素数量。

理解了MNIST，然后再来看[zalandoresearch/fashion-mnist: A MNIST-like fashion product database. Benchmark](https://github.com/zalandoresearch/fashion-mnist)就非常容易理解了。

## 实践

一般的解析思路就是首先获取meta数据，magic number和dimensions，然后一次性读取data，然后根据dimensions做一次reshape就得到了narray，然后通过opencv imshow一下就能看到图片了。

```py
>>> import numpy as np
>>> x = np.arange(6)
>>> x
array([0, 1, 2, 3, 4, 5])
>>> x.reshape((3,2))
array([[0, 1],
       [2, 3],
       [4, 5]])
```

### labels
```py
#!/usr/bin/env python3

import struct
import numpy as np

# labels
with open('./data/FashionMNIST/raw/t10k-labels-idx1-ubyte', 'rb') as f:
  magic_number = int.from_bytes(f.read(4), 'big')
  print(f"magic number: {magic_number:#010x}")

  # 简化版本，未分析mgic_number中第3个字节
  label_count = int.from_bytes(f.read(4), 'big')
  print(f"label count: {label_count}")

  label_data = f.read()
  labels = np.frombuffer(label_data, dtype=np.uint8)
  print(labels)
```

Outputs
```
magic number: 0x00000801
label count: 10000
[9 2 1 ... 8 1 5]
```

### images

```py
#!/usr/bin/env python3

import numpy as np
import cv2

def get_images():
  with open('./data/FashionMNIST/raw/t10k-images-idx3-ubyte', 'rb') as f:
    magic_number = int.from_bytes(f.read(4), 'big')
    print(f"magic number: {magic_number:#010x}")

    image_count = int.from_bytes(f.read(4), 'big')
    row_count = int.from_bytes(f.read(4), 'big')
    column_count = int.from_bytes(f.read(4), 'big')
    print(f"image count: {image_count}")
    print(f"row count: {row_count}")
    print(f"column count: {column_count}")

    image_data = f.read()
    images = np.frombuffer(image_data, dtype=np.uint8).reshape((image_count, row_count, column_count))
    return images

if __name__ == "__main__":
  image = get_images()[1]
  print(image)
  cv2.imshow('img', image)
  cv2.waitKey()
  cv2.destroyAllWindows()
```

- 需要通过`pip install opencv-python`, conda install会找不到，估计需要添加相应的channel

然后用我蹩脚的python写了一下拼接，几行代码就可以通过np实现下面的效果

```py
if __name__ == "__main__":
  images = get_images()

  lines = []
  for i in range(50):
    line = images[i*50]
    for j in range(49):
      line = np.concatenate((line, images[i*50+j+1]), axis=1)
    lines.append(line)

  result = lines[0]
  for i in range(10):
    result = np.concatenate((result, lines[i+1]), axis=0)

  cv2.imshow('img', result)
  cv2.waitKey()
  cv2.destroyAllWindows()
```

<img src="https://user-images.githubusercontent.com/1457904/152578887-84f88f4f-79b0-41a0-b856-c89d3a1335f7.png">

matplotlib version

```py
if __name__ == "__main__":
  label_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat', 'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']
  labels = get_labels()
  images = get_images()

  fig = plt.figure()
  for i in range(30):
    plt.subplot(5, 6, i+1)
    plt.tight_layout()
    plt.imshow(images[i], cmap='gray', interpolation='none')
    plt.title(label_names[labels[i]])
    plt.xticks([])
    plt.yticks([])

  plt.show()
```

<img src="https://user-images.githubusercontent.com/1457904/152580603-01a9d691-eb8a-4f0b-801b-358a45934710.png">

# 参考阅读:
- [从0手撕MNIST - 知乎](https://zhuanlan.zhihu.com/p/424152056)
